File: kernel/src/main/kotlin/ninja/blacknet/transaction/Lease.kt
/*
 * Copyright (c) 2018-2024 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import kotlinx.serialization.Serializable
import ninja.blacknet.core.*
import ninja.blacknet.crypto.Hash
import ninja.blacknet.crypto.PoS
import ninja.blacknet.crypto.PublicKey
import ninja.blacknet.serialization.LongSerializer

/**
 * 創建合約
 */
@Serializable
class Lease(
        @Serializable(with = LongSerializer::class)
        val amount: Long,
        val to: PublicKey
) : TxData {
    override fun processCoinImpl(tx: Transaction, hash: Hash, dataIndex: Int, coinTx: CoinTx): Status {
        if (amount < PoS.MIN_LEASE) {
            return Invalid("$amount less than minimal ${PoS.MIN_LEASE}")
        }
        val account = coinTx.getAccount(tx.from)!!
        val status = account.credit(amount)
        if (status != Accepted) {
            return status
        }
        coinTx.setAccount(tx.from, account)
        val toAccount = coinTx.getOrCreate(to)
        toAccount.leases.add(AccountState.Lease(tx.from, coinTx.height(), amount))
        coinTx.setAccount(to, toAccount)
        return Accepted
    }

    fun involves(publicKey: PublicKey) = to == publicKey
}

File: kernel/src/main/rust/transaction/lease.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::amount::Amount;
use crate::blake2b::Hash;
use crate::ed25519::PublicKey;
use crate::error::{Error, Result};
use crate::proofofstake::MIN_LEASE;
use crate::transaction::{CoinTx, Transaction, TxData};
use alloc::format;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
pub struct Lease {
    amount: Amount,
    to: PublicKey,
}

impl Lease {
    pub const fn new(amount: Amount, to: PublicKey) -> Self {
        Self { amount, to }
    }

    pub const fn amount(&self) -> Amount {
        self.amount
    }

    pub const fn to(&self) -> PublicKey {
        self.to
    }
}

impl TxData for Lease {
    fn process_impl(
        &self,
        tx: Transaction,
        _hash: Hash,
        _data_index: u32,
        coin_tx: impl CoinTx,
    ) -> Result<()> {
        if self.amount < MIN_LEASE {
            return Err(Error::Invalid(format!(
                "{0} less than minimal {MIN_LEASE}",
                self.amount
            )));
        }
        let mut account = coin_tx.get_account(tx.from())?;
        account.credit(self.amount)?;
        coin_tx.set_account(tx.from(), account);
        let mut to_account = coin_tx.get_or_create(self.to);
        to_account.add_lease(tx.from(), coin_tx.height(), self.amount);
        coin_tx.set_account(self.to, to_account);
        Ok(())
    }
}
